src/remoteClient/clientContextImpl.cpp

```diff
@@ -4351,8 +4385,18 @@ class InternalClientContextImpl :
         BeaconHandler::shared_pointer handler;
         if (it == m_beaconHandlers.end())
         {
+            // If we're tracking too many beacons, we'll just ignore this one
+            if (m_beaconHandlers.size() >= maxTrackedBeacons)
+            {
+                char ipa[64];
+                sockAddrToDottedIP(&responseFrom->sa, ipa, sizeof(ipa));
+                LOG(logLevelDebug, "Tracked beacon limit reached (%d), ignoring %s\n", maxTrackedBeacons, ipa);
+                return BeaconHandler::shared_pointer();
+            }
+
             // stores weak_ptr
             handler.reset(new BeaconHandler(internal_from_this(), responseFrom));
+            m_timer->scheduleAfterDelay(BeaconCleanupCallback::shared_pointer(new BeaconCleanupCallback(*this, *responseFrom)), maxBeaconLifetime);
```

Comment:
`class Timer` can be a little strange to use.  As written, this timer will never be `cancel()`ed.  If fact, since the callback pointer isn't be stored, it can't be canceled.  So each time a new server appears, this timer will drop it 360 seconds later.  Then, since that server will probably still be active, it will be "new" again 180 after than.

I think you are on the right track though.  What you would need to do is to store a `BeaconCleanupCallback` in (or as part of) `BeaconHandler`.  Then the logic would be to start the timer on creation (first beacon).  Then cancel and restart it on each subsequent beacon.  This way the timer will only expire when the corresponding server stops sending beacons.